test/unit/query_expr_compiler/test_rewrite_rules.py

```diff
@@ -0,0 +1,520 @@
+"""Tests for rewrite rules."""
+from dataclasses import replace
+from typing import Optional, Union
+
+import pytest
+from tmlt.core.domains.spark_domains import (
+    SparkColumnDescriptor,
+    SparkFloatColumnDescriptor,
+    SparkIntegerColumnDescriptor,
+)
+from tmlt.core.measurements.aggregations import NoiseMechanism
+from tmlt.core.measures import ApproxDP, PureDP, RhoZCDP
+
+from tmlt.analytics import KeySet
+from tmlt.analytics._catalog import Catalog
+from tmlt.analytics._query_expr import (
+    AverageMechanism,
+    CountDistinctMechanism,
+    CountMechanism,
+    GroupByBoundedAverage,
+    GroupByBoundedSTDEV,
+    GroupByBoundedSum,
+    GroupByBoundedVariance,
+    GroupByCount,
+    GroupByCountDistinct,
+    PrivateSource,
+    QueryExpr,
+    StdevMechanism,
+    SumMechanism,
+    SuppressAggregates,
+    VarianceMechanism,
+)
+from tmlt.analytics._query_expr_compiler._rewrite_rules import CompilationInfo, rewrite
+from tmlt.analytics._schema import ColumnDescriptor, ColumnType
+
+# SPDX-License-Identifier: Apache-2.0
+# Copyright Tumult Labs 2025
+
+
+@pytest.fixture(name="test_data", scope="class")
+def setup_test(request):
+    """Setup tests."""
+    catalog = Catalog()
+    catalog.add_private_table(
+        "private",
+        {
+            "A": ColumnDescriptor(ColumnType.VARCHAR),
+            "B": ColumnDescriptor(ColumnType.INTEGER),
+            "X": ColumnDescriptor(ColumnType.DECIMAL),
+        },
+    )
+    request.cls.catalog = catalog
+
+
+BASE_EXPR = PrivateSource("private")
+
+
+@pytest.mark.usefixtures("test_data")
+class TestRewriteRules:
+    """Tests for rewrite rules."""
+
+    catalog: Catalog
+    base_expr: QueryExpr
+
+    @pytest.mark.parametrize(
+        "query_mechanism,output_measure,expected_mechanism",
+        [
+            (mech, meas, NoiseMechanism.GEOMETRIC)
+            for mech in [CountMechanism.DEFAULT, CountMechanism.LAPLACE]
+            for meas in [PureDP(), ApproxDP()]
+        ]
+        + [
+            (mech, RhoZCDP(), NoiseMechanism.DISCRETE_GAUSSIAN)
+            for mech in [CountMechanism.DEFAULT, CountMechanism.GAUSSIAN]
+        ]
+        + [
+            (CountMechanism.LAPLACE, meas, NoiseMechanism.GEOMETRIC)
+            for meas in [PureDP(), ApproxDP(), RhoZCDP()]
+        ],
+    )
+    def test_noise_selection_for_count(
+        self,
+        query_mechanism: CountMechanism,
+        output_measure: Union[PureDP, RhoZCDP],
+        expected_mechanism: NoiseMechanism,
+    ) -> None:
+        """Test noise selection for GroupByCount query expressions."""
+        expr = GroupByCount(
+            child=BASE_EXPR,
+            groupby_keys=KeySet.from_dict({}),
+            mechanism=query_mechanism,
+        )
+        info = CompilationInfo(output_measure=output_measure, catalog=self.catalog)
+        got_expr = rewrite(info, expr)
```

Comment:
This is fine for now, but I'm concerned that as we add more rewrite rules, this will start breaking. Maybe it might make more sense to only call the specific rewrite rule we want to test?